### PR TITLE
Add ECDSA support

### LINKED ISSUE #8: Add ECDSA support

### PR DESCRIPTION
**Pull request by elichai**
Closes https://github.com/kaspanet/go-secp256k1/issues/8

### CODE REVIEW

File: ecdsa_privkey.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+// #include "./depend/secp256k1/include/secp256k1.h"
+import "C"
+import (
+	"crypto/rand"
+	"github.com/pkg/errors"
+	"unsafe"
+)
+
+// ECDSAPrivateKey is a type representing a Secp256k1 private key.
+// This private key can be used to create Schnorr/ECDSA signatures
```
Comment (someone235) [inline ecdsa_privkey.go]: So why is it called ECDSA?

File: ecdsa_privkey.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	return &pubkey, nil
+}
+
+// ECDSASign creates an ECDSA signature using the private key and the input hashed message.
+// Notice: the [32] byte array *MUST* be a hash of a message.
+func (key *ECDSAPrivateKey) ECDSASign(hash *Hash) (*ECDSASignature, error) {
+	var auxilaryRand [32]byte
+	n, err := rand.Read(auxilaryRand[:])
+	if err != nil || n != len(auxilaryRand) {
+		return nil, err
```
Comment (stasatdaglabs) [inline ecdsa_privkey.go:37]: This branch returns (nil, nil) if `n != len(auxilaryRand)`. Please split into two conditions instead.

File: ecdsa_privkey.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+// GenerateECDSAPrivateKey generates a random valid private key from `crypto/rand`
+func GenerateECDSAPrivateKey() (key *ECDSAPrivateKey, err error) {
+	key = &ECDSAPrivateKey{init: true}
+	cPtr := (*C.uchar)(&key.privateKey[0])
+	for {
+		n, tmpErr := rand.Read(key.privateKey[:])
+		if tmpErr != nil || n != len(key.privateKey) {
+			return nil, tmpErr
```
Comment (stasatdaglabs) [inline ecdsa_privkey.go:95]: This branch returns (nil, nil) if `n != len(key.privateKey)`. Please split into two conditions instead.

File: ecdsa_privkey.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+		}
+		ret := C.secp256k1_ec_seckey_verify(C.secp256k1_context_no_precomp, cPtr)
+		if ret == 1 {
+			return
```
Comment (stasatdaglabs) [inline ecdsa_privkey.go]: For readability's sake, please avoid implicit return values.

File: ecdsa_privkey.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+		}
+	}
+}
+
+// Serialize a private key
+func (key *ECDSAPrivateKey) Serialize() *SerializedPrivateKey {
+	ret := SerializedPrivateKey(key.privateKey)
```
Comment (stasatdaglabs) [inline ecdsa_privkey.go:109]: Please consider making sure that the private key is initialized.

**Review by someone235 — CHANGES_REQUESTED** [review]

File: ecdsa_pubkey.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+// ECDSAVerify verifies a ECDSA signature using the public key and the input hashed message.
+// Notice: the [32] byte array *MUST* be a hash of a message you hashed yourself.
+func (key *ECDSAPublicKey) ECDSAVerify(hash *Hash, signature *ECDSASignature) bool {
+	cPtrHash := (*C.uchar)(&hash[0])
+	return C.secp256k1_ecdsa_verify(context, &signature.signature, cPtrHash, &key.pubkey) == 1
+}
+
+// DeserializeECDSAPubKey deserializes a serialized ECDSA public key, verifying it's valid.
+// it supports both compressed(33 bytes) and uncompressed(65 bytes) public keys.
```
Comment (stasatdaglabs) [inline ecdsa_pubkey.go]: The code appears to error if give an uncompressed public key.

File: ecdsa_privkey.go (hunk 1 above)
Comment (elichai) [reply ecdsa_privkey.go]: This is wrong :)

**Review by stasatdaglabs — CHANGES_REQUESTED** [review]

File: ecdsa_privkey.go (hunk 2 above)
Comment (elichai) [reply ecdsa_privkey.go:37]: https://golang.org/pkg/crypto/rand/#Read
>  Read is a helper function that calls Reader.Read using io.ReadFull. On return, n == len(b) if and only if err == nil.

File: ecdsa_pubkey.go (hunk 6 above)
Comment (elichai) [reply ecdsa_pubkey.go]: I'll edit the comment

**Review by stasatdaglabs — APPROVED** [review]

**Review by someone235 — APPROVED** [review]

**Pull request merged**